### PR TITLE
[MM-19914] Fix data races in WebSocketEvent

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR fixes several data race conditions happening in `model.WebSocketEvent` especially during JSON serialization/de-serialization.

The proposed fix changes `model.WebSocketEvent` to be immutable by:

- Making its fields unexported.
- Providing getters/setters. 
- Returning a shallow copy every time a change is made.

This way we avoid the use of a locking mechanism (e.g. `sync.Mutex`) and at the same time eliminate the concurrent writes which are causing the data races.

#### Ticket

https://mattermost.atlassian.net/browse/MM-19914

### CODE REVIEW

File: model/websocket_message.go [hunk 1]
```diff
@@ -74,57 +74,125 @@ type precomputedWebSocketEventJSON struct {
 	Broadcast json.RawMessage
 }
 
-type WebSocketEvent struct {
```
Comment (hanzei) [inline model/websocket_message.go]: Even tough I think these changes are very valuable, they are breaking semver and hence we can't do this without releasing a new major version.

Comment (streamer45) [reply model/websocket_message.go]: Thank you @hanzei , that's a good point. The type name itself did not change, nor the existing functions but I agree that making the fields unexported and replacing them with getters can be considered a breaking change. Not sure there's a robust way to fix this issue without making that type [opaque](https://en.wikipedia.org/wiki/Opaque_data_type).

Comment (hanzei) [reply model/websocket_message.go]: I can't think of a way either. I would opt for exporting the fields, but marking them as deprecated and recommend to use the Getter.

Comment (agnivade) [reply model/websocket_message.go]: Yep, let's export them and mark as deprecated. Great work @streamer45

Comment (streamer45) [reply model/websocket_message.go]: Sounds good to me. @jespino, @jwilander thoughts?

Comment (agnivade) [reply model/websocket_message.go]: And also to add a comment to unexport the fields when 6.x comes out.

Comment (jespino) [reply model/websocket_message.go]: Yes, I think is enough to have them published, my concern would be around miss-usage of them generating race conditions, but I guess it was racy before.

**Comment by jasonblais** [comment]
@ogi-m Quick reminder to help with QA review for this PR. Thanks! :)

**Comment by streamer45** [comment]
@agnivade , @hanzei last chance to take a last look at this before we merge.

**Comment by agnivade** [comment]
Oh, there is nothing blocking this review from my side. It is good to go. :+1:

**Comment by hanzei** [comment]
I think it's good to go too

**Pull request merged**